snapcraft/internal/meta/_snap_packaging.py

```diff
@@ -605,9 +610,40 @@ def _wrap_apps(self, apps: Dict[str, Any]) -> Dict[str, Any]:
             adapter = Adapter[app.pop("adapter").upper()]
             if adapter == Adapter.LEGACY:
                 self._wrap_app(app_name, app)
+            elif adapter == Adapter.FULL:
+                self._add_command_chain_to_app(app_name, app)
             self._generate_desktop_file(app_name, app)
         return apps
 
+    def _add_command_chain_to_app(self, app_name, app):
+        # First, validate the commands. Without a wrapper, we're bound by snapd's
+        # restrictions. No quotes, and the first string must be a relative path from
+        # the root of the snap.
+        commands = (app[k] for k in ("command", "stop-command") if k in app)
+        for command in commands:
+            if not _APP_COMMAND_PATTERN.match(command):
+                raise errors.InvalidAppCommandFormatError(command, app_name)
+
+            command_without_args = command.split()[0]
+            binary_path = os.path.join(self._prime_dir, command_without_args)
+            is_executable = False
+            with contextlib.suppress(FileNotFoundError):
+                mode = os.stat(binary_path).st_mode
+                is_executable = (
+                    mode & stat.S_IXUSR or mode & stat.S_IXGRP or mode & stat.S_IXOTH
+                )
+
+            if not is_executable:
+                raise errors.InvalidAppCommandError(command_without_args, app_name)
+
+        # Finally, assuming there is a meta runner, add it to the BEGINNING of the
+        # command chain. This is to ensure all subsequent commands in the chain get
+        # a valid environment.
+        if os.path.isfile(self._meta_runner):
```

Comment:
this made me scratch my head a bit in the sense that we might not have picked the right name. meta is very meta :-)

Reply:
Haha, yeah it's not the best. The naming challenge comes from the fact that eventually plugins will be supplying command chain binaries as well, so we need to differentiate the "base" one coming from meta, so I went with... meta. I'm open to suggestions.